lib/smart_proxy_remote_execution_ssh/runners/script_runner.rb

```diff
@@ -291,34 +287,10 @@ def settings
     # available. The yielding doesn't happen automatically, but as
     # part of calling the `refresh` method.
     def run_async(command)
-      raise 'Async command already in progress' if @started
-
-      @started = false
-      @user_method.reset
-
-      session.open_channel do |channel|
-        channel.request_pty
-        channel.on_data do |ch, data|
-          publish_data(data, 'stdout') unless @user_method.filter_password?(data)
-          @user_method.on_data(data, ch)
-        end
-        channel.on_extended_data { |ch, type, data| publish_data(data, 'stderr') }
-        # standard exit of the command
-        channel.on_request('exit-status') { |ch, data| publish_exit_status(data.read_long) }
-        # on signal: sending the signal value (such as 'TERM')
-        channel.on_request('exit-signal') do |ch, data|
-          publish_exit_status(data.read_string)
-          ch.close
-          # wait for the channel to finish so that we know at the end
-          # that the session is inactive
-          ch.wait
-        end
-        channel.exec(command) do |_, success|
-          @started = true
-          raise('Error initializing command') unless success
-        end
-      end
-      session.process(0) { !run_started? }
+      @command_out, slave = PTY.open
+      @command_in, write = IO.pipe
+      @command_pid = spawn({'SSHPASS' => ssh_options[:password]}, '/usr/bin/sshpass', '-e', '/usr/bin/ssh', @host, *options, command, :in => @command_in, :out => slave)
+      @command_in.close
```

Comment:
This is roughly the same in both `run_sync` and `run_async`. Could we extract it into a method which would open the connection and return all the relevant handles?